Fixes #8151 Build flavors for Amazon Appstore and Google Playstore

mrudultora opened the pull request:
## Purpose / Description
Different flavors added for Amazon App store and Google Play store _**(we cannot use default here)**_. And, I tried to hide the camera related options in amazon flavor. 

## Fixes
Fixes #8151 

## Approach
Added the flavors in gradle. Created a class to get all the permissions to find out whether camera permission is present in manifest or not. If not present. hide the camera related options from menu's.
**4 Build variants are generated now (amazonDebug, amazonRelease, playDebug, playRelease)**

![img](https://user-images.githubusercontent.com/65870389/111910471-80bcb180-8a87-11eb-926c-c0cdeb485519.png)


## How Has This Been Tested?
Tested on real device (Redmi Note 7S and API 29).

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+/***************************************************************************************
+ *                                                                                      *
+ * Copyright (c) 2012 Norbert Nagold <norbert.nagold@gmail.com>                         *
```
Is this copyright header correct?

mrudultora replied:
I think yes. Would you please elaborate?

david-allison-1 replied:
You don't have your name/email in the header, and you've included past contributors

mrudultora replied:
Ohh ! Actually I thought I am a new contributor so not allowed to use our name. I will correct that... ;-)

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+public class CheckCameraPermission {
+    private final Context mContext;
+
+    public CheckCameraPermission(Context mContext) {
```
param name: `context`

mrudultora replied:
Would you tell me where we have to use variable starting with letter **_m_**. It confuses me ...

david-allison-1 replied:
Code style: [Non-public, non-static field names should start with m](https://github.com/ankidroid/Anki-Android/wiki/Code-style#non-public-non-static-field-names-should-start-with-m)

Note, that I wouldn't apply this rule for any new projects, but consistency is most important.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java:
```diff
@@ -74,6 +75,9 @@
     private int mFieldIndex;
 
+    /* To check whether Camera Permission is asked in AndroidManifest.xml */
+    private CheckCameraPermission mCheckCameraPermission;
```
You should be able to initialise the variable inline (here, and in the other class)

mrudultora replied:
Ya, I will do that.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+        this.mContext = mContext;
+    }
+
+    public boolean checkManifestCameraPermission() {
```
This seems to be the wrong way around - from the name, I'd expect a return value of true if the camera is supported.

Maybe rename the method: `cameraPermissionDisabled`

mrudultora replied:
Nope, this method is used to check whether the manifest contains camera permission. I think the name you suggested is not suitable. Any different name and more specific name ?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+                return false;
+            }
+        } catch (PackageManager.NameNotFoundException e) {
+            e.printStackTrace();
```
```suggestion
            Timber.w(e);
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+            if (Arrays.toString(requestedPermissions).contains("android.permission.CAMERA")) {
+                return false;
+            }
+        } catch (PackageManager.NameNotFoundException e) {
```
I'd be tempted to catch a general exception here

mrudultora replied:
Any specific reason?

david-allison-1 replied:
We don't want to leak an exception if it worked before, and package permissions feels like there's a lot of scope for unknown unknowns in terms of exceptions

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/CheckCameraPermission.java:
```diff
@@ -0,0 +1,46 @@
+                return false;
+            }
+        } catch (PackageManager.NameNotFoundException e) {
+            e.printStackTrace();
```
If an exception is thrown, what do we want the default to be?

mrudultora replied:
It will return true, i.e. if we are on fire devices it will show a toast (not much problem here I guess).

david-allison-1 reviewed (commented):
I'm not qualified to review this. Only nitpicks

My main question would be whether there is a way to keep the camera permissions for Fire devices which do have a camera.

david-allison-1 commented:
Ci is failing

```
* What went wrong:
Task 'compileReleaseJavaWithJavac' not found in project ':AnkiDroid'.
```

mrudultora commented:
> I'm not qualified to review this. Only nitpicks
> 
> My main question would be whether there is a way to keep the camera permissions for Fire devices which do have a camera.

No, actually the problem is Amazon is not allowing us to upload if we have Camera permission. As mike suggested we have to remove this permission. Then only it is possible to upload.

mrudultora commented:
> Ci is failing
> 
> ```
> * What went wrong:
> Task 'compileReleaseJavaWithJavac' not found in project ':AnkiDroid'.
> ```

May be due to the flavors.

mrudultora commented:
@david-allison-1 Changes done. Also, if you know any other menu or place (other than one I resolved here ) where camera permissions are being asked, could you tell me? Because then I have to do some changes there also ;-)

david-allison-1 commented:
You got them all (just the one). 

You should be able to find references on this: https://github.com/ankidroid/Anki-Android/blob/9892bea7061ff97a05f266ddea4ca63e18d48bec/AnkiDroid/src/main/java/com/ichi2/utils/Permissions.java#L13-L16

mrudultora commented:
@david-allison-1 The problem in using this method is .... Let's assume you are using **playDebug** flavor. Now, initially we have not granted the permissions so the option to add image will not at all come in pop-up menu. But it must come there because it's a play store variant ( this method will work fine for amazon flavor though). The method I used is somewhat reliable. We are looking for the permission in manifest. If it's there then shows the option in menu's or else hide them.

david-allison-1 commented:
> @david-allison-1 The problem in using this method is .... Let's assume you are using **playDebug** flavor. Now, initially we have not granted the permissions so the option to add image will not at all come in pop-up menu. But it must come there because it's a play store variant ( this method will work fine for amazon flavor though). The method I used is somewhat reliable. We are looking for the permission in manifest. If it's there then shows the option (whether permission is granted or not) in menu's or else hide them.

Oh no - just as an example of where the app should be using the Camera - all seems good from that respect

mrudultora commented:
@mikehardy Your reviews on this PR ?

mikehardy commented on AnkiDroid/src/main/AndroidManifest.xml:
```diff
@@ -37,7 +37,6 @@
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.RECORD_AUDIO" />
     <uses-permission android:name="android.permission.WAKE_LOCK" tools:node="replace"/>
-    <uses-permission android:name="android.permission.CAMERA" />
```
if the Amazon flavor removes the node, the main manifest should keep the node shouldn't it?

mrudultora replied:
No, we don't want any Camera permission in Amazon flavor. So, if present in any manifest it will be added in manifest merging. So, I think its correct to remove it from main manifest also.

mikehardy commented on AnkiDroid/src/play/AndroidManifest.xml:
```diff
@@ -0,0 +1,7 @@
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:installLocation="auto">
+
+    <uses-permission android:name="android.permission.CAMERA" />
```
If the amazon flavor removes it, and the main one has it, I think we might be able to get away with no special manifest for the play flavor at all, at which point I would actually rename the "play" variant to "default" in the build.gradle

mrudultora replied:
Ya! I will change that.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java:
```diff
@@ -215,6 +216,12 @@ public boolean onCreateOptionsMenu(Menu menu) {
         menu.findItem(R.id.multimedia_edit_field_to_image).setVisible(mField.getType() != EFieldType.IMAGE);
+
+        /* To check whether Camera Permission is asked in AndroidManifest.xml */
+        CheckCameraPermission mCheckCameraPermission = new CheckCameraPermission(this);
```
it seems like we are not actually doing anything with this member variable. Could the CheckCameraPermission object just have a static method that verifies the permission?

I'd name it differently as well :-), `CameraPermissionCheck.hasPermission(Context)` seems like it would work

david-allison-1 replied:
(optional) maybe `isPermissionAvailable` - to differentiate it from `Permissions.canUseCamera()`

mikehardy replied:
@david-allison-1's suggested name is much better than mine, I didn't realize how ambiguous mine was until he mentioned it

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -1626,9 +1626,18 @@ private void setMMButtonListener(ImageButton mediaButton, final int index) {
+
+                if (mCheckCameraPermission.checkManifestCameraPermission()) {
+                    MenuItem item = popup.getMenu().findItem(R.id.menu_multimedia_photo);
+                    item.setVisible(false);
```
this is a nice touch

mikehardy reviewed (changes requested):
Thanks for your patience @mrudultora I just today caught up on notifications from github and this was not such a high priority since we don't have that many amazon users and there is still a valid build out there for them

I think it can be simplified a fair bit but I also think it could work well in this style, after you look at the changes I'll try pulling it locally and building from it to see what the .apks look like

mrudultora commented:
As Mike suggested, I added Camera Permission in our main manifest. And also removed the AndroidManifest for **play** flavor. Renaming **play** flavor to **default** is not possible as this keyword is not allowed in gradle for naming purposes. Also changed the method name. It's working as intended. Tested both the flavors on real device.

mrudultora commented:
@david-allison-1 I am just messed up with git rebase. Would it be ok if I create a new PR? Actually, this PR right now is completely okay, I just forgot to change the exception handling part (catching a **Exception**). And, after resolving the merge conflicts there were 93 commits in the pull and I just messed rebase.

david-allison-1 commented:
Don't worry! Getting out of a bad rebase is a skill, but come into advanced git operations gradually, it'll take time

mikehardy commented:
Ha, I've done that (entered rebase hell). As David mentioned it may be worth it - just to play with it! - to try to fix it. Sometimes a a new branch clean from main branch then cherry-pick the stuff you want on to it works easily. Sometimes you can `git reset --soft HEAD~1` to peel the 1st commit off (or more than 1 if more...) then git reset so they are unstaged, then git stash push to store them to the side, then `git --reset hard upstream/master` to clean everything out then re-create the commits etc from git stash pop (or maybe cherry-pick from the old hashes)

The important thing I always remember is that each entry is really just a text diff at the end of the day, which is actually just a blob of text. You can peel text blobs off and put them anywhere or stack them on anything or remove other text blobs - when I think of it like that it just seems like it shouldn't be that bad and I can do it and I get through it fine.
